lib/dpl/provider/bintray.rb

```diff
@@ -0,0 +1,505 @@
+require 'json'
+require 'net/http'
+require "uri"
+require 'find'
+
+module DPL
+  class Provider
+    class Bintray < Provider
+      @test_mode = false
+
+      def check_auth
+      end
+
+      def needs_key?
+        false
+      end
+
+      attr_writer :test_mode
+
+      def init_from_args
+        @user = options[:user]
+        @key = options[:key]
+        @url = options[:url]
+        @file = options[:file]
+        @passphrase = options[:passphrase]
+        @dry_run = options[:dry_run]
+
+        if @user.nil?
+          abort("The 'user' argument is required")
+        end
+        if @key.nil?
+          abort("The 'key' argument is required")
+        end
+        if @file.nil?
+          abort("The 'file' argument is required")
+        end
+        if @url.nil?
+          @url = 'https://api.bintray.com'
+        end
+        if @dry_run.nil?
+          @dry_run = false
+        end
+      end
+
+      def read_descriptor
+        log "Reading descriptor file: #{@file}"
+        @descriptor = JSON.parse(File.read(@file))
+      end
+
+      def descriptor=(json)
+        @descriptor = JSON.parse(json)
+      end
+
+      def head_request(path)
+        url = URI.parse(@url)
+        req = Net::HTTP::Head.new(path)
+        req.basic_auth @user, @key
+
+        sock = Net::HTTP.new(url.host, url.port)
+        sock.use_ssl = true
+        res = sock.start {|http| http.request(req) }
+
+        return res
+      end
+
+      def post_request(path, body)
+        req = Net::HTTP::Post.new(path)
+        req.add_field('Content-Type', 'application/json')
+        req.basic_auth @user, @key
+        if !body.nil?
+          req.body = body.to_json
+        end
+
+        url = URI.parse(@url)
+        sock = Net::HTTP.new(url.host, url.port)
+        sock.use_ssl = true
+        res = sock.start {|http| http.request(req) }
+        return res
+      end
+
+      def put_file_request(local_file_path, upload_path, matrix_params)
+        url = URI.parse(@url)
+
+        data = File.read(local_file_path)
+        http = Net::HTTP.new(url.host, url.port)
+        http.use_ssl = true
+
+        params = ''
+        if !matrix_params.nil?
+          matrix_params.each do |key, val|
+            params << ";#{key}=#{val}"
+          end
+          upload_path << params
+        end
+
+        request = Net::HTTP::Put.new("#{upload_path}")
+        request.basic_auth @user, @key
+        request.body = data
+
+        return http.request(request)
+      end
+
+      def upload_file(artifact)
+        log "Uploading file '#{artifact.local_path}' to #{artifact.upload_path}"
+
+        package = @descriptor["package"]
+        version = @descriptor["version"]
+        package_name = package["name"]
+        subject = package["subject"]
+        repo = package["repo"]
+        version_name = version["name"]
+
+        path = "/content/#{subject}/#{repo}/#{package_name}/#{version_name}/#{artifact.upload_path}"
+        if !@dry_run
```

Comment:
This is a bit strange place to check the value of `@dry_run`. There is nothing this method should do if `@dry_run` is truth-y, so it should be checked up front:

```ruby
def upload_file(artifact)
  return unless @dry_run
  ⋮
end
```

(Having a method named `dry_run?` would be more Ruby-like.)